src/main/java/com/zendesk/maxwell/MaxwellAbstractRowsEvent.java

```diff
@@ -275,21 +292,59 @@ public Object getData(String string) {
 		return list;
 	}
 
-	public List<JSONObject> toJSONObjects() {
-		ArrayList<JSONObject> list = new ArrayList<>();
+	static final JsonFactory factory = new JsonFactory();
 
-		for ( RowMap map : jsonMaps() ) {
-			list.add(new MaxwellJSONObject(map));
+	private JsonGenerator createJSONGenerator(ByteArrayOutputStream b) {
+		try {
+			return factory.createGenerator(b);
+		} catch (IOException e) {
 		}
-		return list;
+		return null;
+	}
+
+	private void rowMapToJSON(JsonGenerator g, RowMap map) throws IOException {
+		g.writeStartObject(); // {
+
+		for ( String key: map.keySet() ) {
+			g.writeObjectField(key, map.get(key)); // type: "insert"
+		}
+
+		g.writeObjectFieldStart("data");
+		for ( String key: map.data.keySet() ) {
+			Object data = map.getData(key);
+
+			if ( data != null ) {
+				if ( data instanceof List ) { // sets come back from .asJSON as lists, and jackson can't deal.
+					List<String> stringList = (List<String>) data;
+
+					g.writeArrayFieldStart(key);
+					for ( String s : stringList )  {
+						g.writeString(s);
+					}
+					g.writeEndArray();
+				} else {
+					g.writeObjectField(key, data);
+				}
+			}
+		}
+		g.writeEndObject();
+		g.writeEndObject();
 	}
 
 	public List<String> toJSONStrings() {
 		ArrayList<String> list = new ArrayList<>();
 
 		for ( RowMap map : jsonMaps() ) {
-			list.add(new MaxwellJSONObject(map).toString());
+			ByteArrayOutputStream b = new ByteArrayOutputStream();
```

Comment:
thanks for the push here.  I had this originally, but the `JsonGenerator` was adding spaces in an annoying way.  Found the work-around and this is worth another 10% or so of overall perf.